File: cn/huohuas001/bot/events/commands/AdministrationCommands.java
package cn.huohuas001.bot.events.commands;

import cn.huohuas001.bot.HuHoBot;
import cn.huohuas001.bot.events.commands.CommandSupport;
import cn.huohuas001.bot.events.commands.Commands;
import cn.huohuas001.bot.state.CommandRepositories;
import cn.huohuas001.bot.provider.HExecution;
import io.github.kloping.qqbot.api.v2.GroupMessageEvent;
import java.util.concurrent.CompletableFuture;
import java.util.function.BiConsumer;

public final class AdministrationCommands
extends CommandSupport {
    @Commands(value={"\u6267\u884c\u547d\u4ee4"})
    public void runServerCommand(HuHoBot huHoBot, GroupMessageEvent groupMessageEvent, String string) {
        if (!this.requireAdmin(huHoBot, groupMessageEvent)) {
            return;
        }
        if (string == null || string.trim().isEmpty()) {
            this.sendDirect(groupMessageEvent, "\u53c2\u6570\u4e0d\u6b63\u786e");
            return;
        }
        this.executeGameCommandWithMention(huHoBot, groupMessageEvent, string, true);
    }

    @Commands(value={"\u7ba1\u7406\u5458\u6267\u884c"})
    public void runAdminCustomCommand(HuHoBot huHoBot, GroupMessageEvent groupMessageEvent, String string) {
        if (!this.requireAdmin(huHoBot, groupMessageEvent)) {
            return;
        }
        if (string == null || string.trim().isEmpty()) {
            this.sendDirect(groupMessageEvent, "\u53c2\u6570\u4e0d\u6b63\u786e");
            return;
        }
        String type = "adminrun";
        String cmd = "huhobot " + type + ' ' + this.groupId(groupMessageEvent) + ' ' + this.userId(groupMessageEvent) + ' ' + string;
        this.executeGameCommandWithMention(huHoBot, groupMessageEvent, cmd, true);
    }

    @Commands(value={"\u5168\u91cf"})
    public void fullAmount(HuHoBot huHoBot, GroupMessageEvent groupMessageEvent, String string) {
        if (!this.requireAdmin(huHoBot, groupMessageEvent)) {
            return;
        }
        String string2 = this.groupId(groupMessageEvent);
        boolean bl = CommandRepositories.INSTANCE.getGroupSettings().fullForwarding(string2, huHoBot.getFullAmount());
        boolean bl2 = !bl;
        CommandRepositories.INSTANCE.getGroupSettings().setFullForwarding(string2, bl2);
        this.sendDirect(groupMessageEvent, "\u672c\u7fa4\u5168\u91cf\u8f6c\u53d1\u5df2" + (bl2 ? "\u5f00\u542f" : "\u5173\u95ed"));
    }

    /**
     * 自定义命令执行 + 带艾特回复（覆盖 CommandSupport.executeGameCommand 的纯文本回复）。
     */
    private void executeGameCommandWithMention(HuHoBot huHoBot, GroupMessageEvent groupMessageEvent, String command, boolean direct) {
        try {
            String outgoingCommand = direct ? command : "huhobot run " + this.groupId(groupMessageEvent) + ' ' + this.userId(groupMessageEvent) + ' ' + command;
            CompletableFuture<HExecution> future = huHoBot.sendCommand(outgoingCommand);
            final HuHoBot fHuHoBot = huHoBot;
            final GroupMessageEvent fEvent = groupMessageEvent;
            future.whenComplete(new BiConsumer<HExecution, Throwable>() {
                @Override
                public void accept(HExecution result, Throwable error) {
                    try {
                        if (error != null || result == null) {
                            sendDirect(fEvent, "\u6e38\u620f\u6865\u63a5\u672a\u914d\u7f6e\u6216\u6267\u884c\u5931\u8d25");
                        } else {
                            String raw = result.getRawString();
                            if (raw == null || raw.trim().isEmpty()) {
                                sendDirect(fEvent, "\u5df2\u53d1\u9001\u6267\u884c\u8bf7\u6c42");
                            } else {
                                sendDirect(fEvent, fHuHoBot.auditText(raw));
                            }
                        }
                    } catch (Throwable t) {
                        // empty catch block
                    }
                }
            });
        } catch (Throwable t) {
            this.sendDirect(groupMessageEvent, "\u6267\u884c\u5931\u8d25: " + t.getMessage());
        }
    }

    private void sendDirect(GroupMessageEvent groupMessageEvent, String string) {
        try {
            String senderName = null;
            try {
                io.github.kloping.qqbot.entities.qqpd.v2.Contact contact = groupMessageEvent.getSender();
                if (contact != null && contact.getUsername() != null) senderName = contact.getUsername();
            } catch (Throwable ignored) {}
            String content = string;
            if (senderName != null && !senderName.isEmpty()) {
                content = "@" + senderName + "\n" + string;
            }
            groupMessageEvent.sendMessage(content);
        }
        catch (Throwable throwable) {
            // empty catch block
        }
    }
}

File: common-Bot/src/main/kotlin/cn/huohuas001/bot/events/commands/AdministrationCommands.kt
package cn.huohuas001.bot.events.commands

import cn.huohuas001.bot.HuHoBot
import cn.huohuas001.bot.QClient
import cn.huohuas001.bot.state.CommandRepositories
import io.github.kloping.qqbot.api.v2.GroupMessageEvent

/**
 * 管理员命令：执行服务器命令、管理员自定义命令、全量转发开关。
 */
class AdministrationCommands : CommandSupport() {

    /** /执行命令 <命令> —— 向服务器控制台派发命令。 */
    @Commands("执行命令")
    fun runServerCommand(plugin: HuHoBot, event: GroupMessageEvent, params: String) {
        if (!requireAdmin(plugin, event)) return
        if (params.trim().isEmpty()) {
            sendDirect(event, "参数不正确")
            return
        }
        executeGameCommandWithMention(plugin, event, params, true)
    }

    /** /管理员执行 <key> —— 以管理员身份执行自定义命令。 */
    @Commands("管理员执行")
    fun runAdminCustomCommand(plugin: HuHoBot, event: GroupMessageEvent, params: String) {
        if (!requireAdmin(plugin, event)) return
        if (params.trim().isEmpty()) {
            sendDirect(event, "参数不正确")
            return
        }
        val cmd = "huhobot adminrun ${groupId(event)} ${userId(event)} $params"
        executeGameCommandWithMention(plugin, event, cmd, true)
    }

    /** /全量 —— 切换本群全量聊天转发。 */
    @Commands("全量")
    fun fullAmount(plugin: HuHoBot, event: GroupMessageEvent, params: String) {
        if (!requireAdmin(plugin, event)) return
        val group = groupId(event)
        val enabled = CommandRepositories.groupSettings.fullForwarding(group, plugin.fullAmount)
        CommandRepositories.groupSettings.setFullForwarding(group, !enabled)
        sendDirect(event, "本群全量转发已${if (!enabled) "开启" else "关闭"}")
    }

    /**
     * 派发命令并带 @ 提及回复执行结果。
     * （覆盖 CommandSupport.executeGameCommand 的纯文本回复，与绑定类命令的回复方式保持一致。）
     */
    private fun executeGameCommandWithMention(plugin: HuHoBot, event: GroupMessageEvent, command: String, direct: Boolean) {
        try {
            val outgoingCommand = if (direct) command
            else "huhobot run ${groupId(event)} ${userId(event)} $command"
            plugin.sendCommand(outgoingCommand).whenComplete { result, error ->
                try {
                    when {
                        error != null || result == null -> sendDirect(event, "游戏桥接未配置或执行失败")
                        result.rawString.isBlank() -> sendDirect(event, "已发送执行请求")
                        else -> sendDirect(event, plugin.auditText(result.rawString))
                    }
                } catch (_: Throwable) {
                }
            }
        } catch (t: Throwable) {
            sendDirect(event, "执行失败: ${t.message}")
        }
    }

    /** 带 @ 提及的 Markdown 回复（失败时退化为纯文本）。 */
    protected fun sendDirect(event: GroupMessageEvent, message: String) {
        try {
            val userId = try {
                userId(event)
            } catch (_: Throwable) {
                null
            }
            val content = if (!userId.isNullOrEmpty() && userId != "<unknown>") "<@$userId>\n$message" else message
            try {
                QClient.replyMarkdown(event, content, null)
            } catch (_: Throwable) {
                event.sendMessage(content)
            }
        } catch (_: Throwable) {
            try {
                event.sendMessage(message)
            } catch (_: Throwable) {
            }
        }
    }
}
